File: Problems/No.703_Kth_Largest_Element_in_a_Stream/kth_largest_element_in_a_stream.py
class KthLargest(object):

    def __init__(self, k, nums):
        """
        :type k: int
        :type nums: List[int]
        """
        self.Kth = k
        self.KminHeap = []

        for i in range(min(self.Kth, len(nums))):
            heapq.heappush(self.KminHeap, nums[i])
        for i in range(self.Kth, len(nums)):
            if (nums[i] > self.KminHeap[0]):
                heapq.heappop(self.KminHeap)
                heapq.heappush(self.KminHeap, nums[i])
        

    def add(self, val):
        """
        :type val: int
        :rtype: int
        """
        if (len(self.KminHeap) < self.Kth):
            heapq.heappush(self.KminHeap, val)
        elif (val > self.KminHeap[0]):
            heapq.heappop(self.KminHeap)
            heapq.heappush(self.KminHeap, val)
        
        return self.KminHeap[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: Problems/No.703_Kth_Largest_Element_in_a_Stream/kth_largest_element_in_a_stream.cpp
class KthLargest {

private:
    int Kth;
    std::priority_queue<int, std::vector<int>, std::greater<int>> KminHeap;

public:
    KthLargest(int k, vector<int>& nums) {
        
        Kth = k;
        for (int i=0; i<std::min(Kth, (int)nums.size()); i++) KminHeap.push(nums[i]);
        for (int i=Kth; i<nums.size(); i++) {
            if (nums[i] > KminHeap.top()) {

                KminHeap.pop();
                KminHeap.push(nums[i]);

            }
        }

    }
    
    int add(int val) {
        
        if (KminHeap.size() < Kth) KminHeap.push(val);
        else if (val > KminHeap.top()) {

            KminHeap.pop();
            KminHeap.push(val);

        }

        return KminHeap.top();

    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */
